sdks/java/extensions/zetasketch/src/test/java/org/apache/beam/sdk/extensions/zetasketch/HllCountTest.java

```diff
@@ -0,0 +1,373 @@
+    STRINGS.forEach(s -> input.add(KV.of("k", s)));
+    PCollection<KV<String, byte[]>> result =
+        p.apply(Create.of(input))
+            .apply(HllCount.Init.stringSketch().withPrecision(TEST_PRECISION).perKey());
+
+    PAssert.that(result)
+        .containsInAnyOrder(Collections.singletonList(KV.of("k", STRINGS_SKETCH_TEST_PRECISION)));
+    p.run();
+  }
+
+  @Test
+  @Category(NeedsRunner.class)
+  public void testInitStringSketchPerKey_WithInvalidPrecision() {
+    List<KV<String, String>> input = new ArrayList<>();
+    STRINGS.forEach(s -> input.add(KV.of("k", s)));
+    thrown.expect(IllegalArgumentException.class);
+    p.apply(Create.of(input)).apply(HllCount.Init.stringSketch().withPrecision(0).perKey());
+  }
+
+  @Test
+  @Category(NeedsRunner.class)
+  public void testInitBytesSketchPerKey() {
+    List<KV<String, byte[]>> input = new ArrayList<>();
+    BYTES.forEach(bs -> input.add(KV.of("k", bs)));
+    PCollection<KV<String, byte[]>> result =
+        p.apply(Create.of(input)).apply(HllCount.Init.bytesSketch().perKey());
+
+    PAssert.that(result).containsInAnyOrder(Collections.singletonList(KV.of("k", BYTES_SKETCH)));
+    p.run();
+  }
+
+  @Test
+  @Category(NeedsRunner.class)
+  public void testMergePartialGlobally() {
+    PCollection<byte[]> result =
+        p.apply(Create.of(INTS1_SKETCH, INTS2_SKETCH)).apply(HllCount.MergePartial.globally());
+
+    PAssert.thatSingleton(result).isEqualTo(INTS1_INTS2_SKETCH);
+    p.run();
+  }
+
+  @Test
+  @Category(NeedsRunner.class)
+  public void testMergePartialGlobally_MergeWithSketchForEmptySet() {
+    PCollection<byte[]> result =
+        p.apply(Create.of(LONGS_SKETCH, LONGS_EMPTY_SKETCH))
+            .apply(HllCount.MergePartial.globally());
+
+    PAssert.thatSingleton(result).isEqualTo(LONGS_SKETCH);
+    p.run();
+  }
+
+  @Test
+  @Category(NeedsRunner.class)
+  public void testMergePartialGlobally_SingletonInput() {
+    PCollection<byte[]> result =
+        p.apply(Create.of(LONGS_SKETCH)).apply(HllCount.MergePartial.globally());
+
+    PAssert.thatSingleton(result).isEqualTo(LONGS_SKETCH);
+    p.run();
+  }
+
+  @Test
+  @Category(NeedsRunner.class)
+  public void testMergePartialGlobally_EmptyInput() {
+    PCollection<byte[]> result =
+        p.apply(Create.empty(TypeDescriptor.of(byte[].class)))
+            .apply(HllCount.MergePartial.globally());
+
+    PAssert.that(result).empty();
+    p.run();
+  }
+
+  @Test
+  @Category(NeedsRunner.class)
+  public void testMergePartialGlobally_IncompatibleSketches() {
+    p.apply(Create.of(INTS1_SKETCH, STRINGS_SKETCH)).apply(HllCount.MergePartial.globally());
+
+    thrown.expect(PipelineExecutionException.class);
+    p.run();
+  }
+
+  @Test
+  @Category(NeedsRunner.class)
+  public void testMergePartialPerKey() {
+    PCollection<KV<String, byte[]>> result =
+        p.apply(
+                Create.of(
+                    KV.of("k1", INTS1_SKETCH),
```

Comment:
I'm wondering whether it's possible that different key has different value types. If that's the case, how can you create different type of sketch from your init.perkey API?

Reply:
Also is it designed that `MergePartial` and `Extract` can work with sketches which are not produced by `Init`? If that's not the case, I would adjust these tests to get sketches from 'Init'.

Reply:
For your first question:
- If you are talking about the input to `HllCount.Init`, it does not seem to be possible because the input type need to be `KV<K, V>` where `V` cannot be different for different keys.
- If you are talking about the input to `HllCount.MergePartial`, whose input is a `KV<K, byte[]>`. It is possible that, e.g. `k1` has a value representing an integer-type sketch, and `k2` has a value representing a string-type sketch. `MergePartial` will work in this case because it does not make assumptions on what the input byte arrays represent.

For your second question:
- Yes that is intended. For example, user can read `byte[]`s from BigQuery to Beam and merge them using `HllCount.MergePartial`.

Reply:
Thanks for explaining! It would be worthy adding these as part of javadoc.

Reply:
Just a quick comment on this: 

> It is possible that, e.g. k1 has a value representing an integer-type sketch, and k2 has a value representing a string-type sketch. MergePartial will work in this case because it does not make assumptions on what the input byte arrays represent.

The (zetasketch HLL++) library will throw an error if you try to merge sketches of different input types. It's important that we do so because the output would be gibberish (e.g., a string-encoded int would probably hash differently than a 'native' int => it would be counted twice in the output).